File: leetcode/199-binary-tree-right-side-view.cpp
/*
Given a binary tree, imagine yourself standing on the right side of it, return the values of the nodes you can see ordered from top to bottom.

For example:
Given the following binary tree,
   1            <---
 /   \
2     3         <---
 \     \
  5     4       <---
You should return [1, 3, 4].

Credits:
Special thanks to @amrsaqr for adding this problem and creating all test cases.
*/

/**
 * Definition for binary tree
 * struct TreeNode {
 *     int val;
 *     TreeNode *left;
 *     TreeNode *right;
 *     TreeNode(int x) : val(x), left(NULL), right(NULL) {}
 * };
 */

struct Node {
    int level;
    TreeNode *node;
    Node(int level, TreeNode *node): level(level), node(node) {}
};

class Solution {
public:
    vector<int> rightSideView(TreeNode *root) {
        vector<int> result;
        queue<Node> q;
        auto currentLevel(0);
        if (root != nullptr) {
            q.push(Node(1, root));
        }
        while (!q.empty()) {
            auto tmp(q.front());
            q.pop();
            if (tmp.level != currentLevel) {
                result.push_back(tmp.node->val);
                currentLevel = tmp.level;
            }
            if (tmp.node->right != nullptr) {
                q.push(Node(tmp.level + 1, tmp.node->right));
            }
            if (tmp.node->left != nullptr) {
                q.push(Node(tmp.level + 1, tmp.node->left));
            }
        }
        return result;
    }
};

File: leetcode/199-binary-tree-right-side-view.rb
# Definition for a binary tree node.
# class TreeNode
#     attr_accessor :val, :left, :right
#     def initialize(val)
#         @val = val
#         @left, @right = nil, nil
#     end
# end

# @param {TreeNode} root
# @return {Integer[]}
def right_side_view(root)
  right_view = []  
  level = [root]
  while not level.empty?
    right_view << level[-1].val unless level[-1].nil?
    level = level.select {|x| not x.nil?}.flat_map {|x| [x.left, x.right]}.select {|x| not x.nil?}
  end
  right_view
end
